### PR TITLE
Detecting inactive desktop session

### PR DESCRIPTION
**Pull request by ajay-arora**
Bug 377167:Handle cases when there is no active session (for ex. when user disconnected from remote desktop without running disconnect.bat)

Problem  : When users connected to remote desktop, but disconnected the session without making the console session active, no session remains active. As a result, if testagent gets configured as a process and UI tests are run that interact with the desktop, the tests will fail.

Fix : During testagent configuration, see if there is any active desktop session. If not found, mark the machine for reboot.

Testing : 
	1. 1st time Run as a service (no reboot, non UI test pass)
	2. 1st time Run as a process
		a. while connected via a remote desktop (no reboot, test pass)
		b. while connected via a remote desktop, and the remote desktop session was minimized  (no reboot, UI test failed. Error message from CUIT layer indicates that either the desktop session was locked or minimized, so thats the experience here.)
		c. while connected via a remote desktop, and the remote desktop session was locked (no reboot, UI test failed. Error message from CUIT layer indicates that either the desktop session was locked or minimized, so thats the experience here.)
		d. while connected via a remote desktop, and the remote desktop session was disconnected via disconnect.bat (no reboot, test pass)
		e. while connected via a remote desktop, and the remote desktop session was disconnected normally without disconnect.bat (reboots, test pass)
		f. while connected via a remote desktop, and the remote desktop session was disconnected via windows -> disconnect (reboots, test pass)
		g. when the machine was never logged into (reboots, test pass)
		h. first time configuration (reboots, test pass)
	3. Back to back runs as a process (no reboot, test pass)
	4. Back to back runs as a service. (no reboot, non UI test pass)
	5. Round trip from Process - Service - Process, when no one had ever logged in to the machine
	6. Round trip from Service - Process - Service, when no one had ever logged in to the machine
	7. Run as a service with invalid params, and TestAgentConfig.exe failed, m/c did not reboot
        8. Run as a process with invalid params, and TestAgentConfig.exe failed, m/c not reboot

Ran all above tests on Win2k8r2. On the rest of the OS, cover only sanity tests (2e)

### CODE REVIEW

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 1]
```diff
@@ -281,47 +281,147 @@ function Set-TestAgentConfiguration
...
+            WTSListen,
+            WTSReset,
+            WTSDown,
+            WTSInit
+        }
+    }
+'@
+
+    $wtsenumsig = @'
+        [DllImport("wtsapi32.dll", SetLastError = true)]
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: any minimum os requirements for these dlls?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 2]
```diff
@@ -620,12 +720,6 @@ function EnableTracing
 
 function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.PSCredential] $MachineCredential)
 {
-    if(IsDtaExecutionHostRunning)
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: why remove this check?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 3]
```diff
@@ -281,47 +281,147 @@ function Set-TestAgentConfiguration
...
         $configArgs = $configArgs +  ("/Capabilities:{0}" -f $Capabilities)
     }
+
     DeleteDTAAgentExecutionService -ServiceName "DTAAgentExecutionService" | Out-Null
 
     $configOut = InvokeTestAgentConfigExe -Arguments $configArgs -Version $TestAgentVersion -UserCredential $MachineUserCredential
 
-    if ($configOut.ExitCode -eq 0 -and $configAsProcess -eq $true)
+    if ($configOut.ExitCode -ne 0 -and $configOut.ExitCode -ne 3010)
     {
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: any reason to not retain the traces?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 4]
```diff
@@ -281,47 +281,147 @@ function Set-TestAgentConfiguration
...
     }
 
-    $doReboot = $false
-    # 3010 is exit code to indicate a reboot is required
-    if ($configOut.ExitCode -eq 3010)
+    if (IsDtaExecutionHostRunning)
     {
-        Write-Verbose -Message "Marking the machine for reboot as exit code 3010 received" -Verbose
-        $doReboot = $true
+        Write-Verbose -Message "Killing the existing instances" -Verbose
```
Comment (allendm-msft) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: Dont use "killing" :) 
Stopping already running instances of DTAExecutionHost

**Comment by allendm-msft** [comment]
signed off with comments

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 (hunk 1 above)
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: yes, its Vista (Client) and Windows Server 2003 (Server)

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 5]
```diff
@@ -281,47 +281,147 @@ function Set-TestAgentConfiguration
     {
         $configArgs = $configArgs +  ("/Capabilities:{0}" -f $Capabilities)
     }
+
     DeleteDTAAgentExecutionService -ServiceName "DTAAgentExecutionService" | Out-Null
 
     $configOut = InvokeTestAgentConfigExe -Arguments $configArgs -Version $TestAgentVersion -UserCredential $MachineUserCredential
 
-    if ($configOut.ExitCode -eq 0 -and $configAsProcess -eq $true)
+    if ($configOut.ExitCode -ne 0 -and $configOut.ExitCode -ne 3010)
```
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: bump up task version

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 6]
```diff
@@ -281,47 +281,147 @@ function Set-TestAgentConfiguration
...
+
+    add-type  $wtssig
+    $wtsenum = add-type -MemberDefinition $wtsenumsig -Name PSWTSEnumerateSessions -Namespace GetLoggedOnUsers -PassThru
+    $wtsOpen = add-type -MemberDefinition $wtsopensig -name PSWTSOpenServer -Namespace GetLoggedOnUsers -PassThru
+    $wtsmessage = Add-Type -MemberDefinition $wtsSendMessagesig -name PSWTSSendMessage -Namespace GetLoggedOnUsers -PassThru
+
+    [long]$count = 0
+    [long]$ppSessionInfo = 0
+
+    $server = $wtsOpen::WTSOpenServer("localhost")
```
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: can u check if this works on azure machines as well

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 (hunk 4 above)
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: i picked up the existing code ... fixed it in both places

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 (hunk 6 above)
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: don't see how its different than a normal machine.

Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: and i would .... but it takes too much effort to get azure scenario working. http based standard env on azure is not supported

**Pull request merged**